### PR TITLE
Blender: Dynamic workflow for Modeling > Rigging > Layout > Animation using linked libraries

### PR DESCRIPTION
**Pull request by jojoba106**
## Brief description
use collections and overrides for the instances.

## Description
change Blender plugins to work with a collection instead of an empty to store custom properties of the avalon instance.
use overrides.
[https://github.com/pypeclub/OpenPype/discussions/2598](https://github.com/pypeclub/OpenPype/discussions/2598)
## Documentation (add _"type: documentation"_ label)
[feature_documentation](future_url_after_it_will_be_merged)

## Testing notes:
1. start with this step
2. follow this step

## To do:
- [ ] Adapt the validate plugins to work with the collection-container logic. For now, they test the collections not the objects
- [x] If there is only one collection in the scene collection, use it to create the instance collection.
- [ ] Merge last model work scene and publish model at the rig step to publish a new model which has been make local

### CODE REVIEW

**Comment by Tilix4** [comment]
Great! Please do:

- [ ] Rename title for: Blender: Dynamic workflow for Modeling > Rigging > Layout > Animation using linked libraries
- [ ] Add link to discussion into description: https://github.com/pypeclub/OpenPype/discussions/2598

**Comment by Tilix4** [comment]
I think it'd be nice to run `black` on your `.py` files to quickly clean basic syntax errors.

File: openpype/hosts/blender/api/plugin.py [hunk 1]
```diff
@@ -49,6 +69,20 @@ def get_unique_number(
...
 
+def get_model_unique_number(current_container_name:str) -> str:
+    """Return a unique number based on the asset name."""
+    data_collections = bpy.data.collections
+    container_names = [c.name for c in data_collections if c.get(AVALON_PROPERTY)]
+    if container_names == []:
+        return "001"
+    count = 1
+    name = f"{current_container_name}_{count:0>3}"
+    while name in container_names:
```
Comment (Tilix4) [inline openpype/hosts/blender/api/plugin.py:80]: Why do you use a `while` instead of a `for` here?

Comment (Tilix4) [inline openpype/hosts/blender/api/plugin.py:80]: Could you please add a comment for this step?

File: openpype/hosts/blender/api/plugin.py [hunk 2]
```diff
@@ -86,6 +120,15 @@ def create_blender_context(active: Optional[bpy.types.Object] = None,
     raise Exception("Could not create a custom Blender context.")
 
 
+def get_instance_list():
+    """Get the parent of the input collection"""
```
Comment (Tilix4) [inline openpype/hosts/blender/api/plugin.py:124]: The doctring doesn't match the concept of `list`.

File: openpype/hosts/blender/api/plugin.py [hunk 3]
```diff
@@ -86,6 +120,15 @@ def create_blender_context(active: Optional[bpy.types.Object] = None,
     raise Exception("Could not create a custom Blender context.")
 
 
+def get_instance_list():
```
Comment (Tilix4) [inline openpype/hosts/blender/api/plugin.py:123]: The function name should be `get_instances_list`. A list must be of "thing**s**"

File: openpype/hosts/blender/plugins/create/create_model.py [hunk 4]
```diff
@@ -21,29 +21,33 @@ def process(self):
...
             selected = lib.get_selection()
             for obj in selected:
-                obj.select_set(True)
-            selected.append(asset_group)
-            bpy.ops.object.parent_set(keep_transform=True)
-
-        return asset_group
+                instance.objects.link(obj)
+                bpy.context.scene.collection.objects.unlink(obj)
+        else:
```
Comment (Tilix4) [inline openpype/hosts/blender/plugins/create/create_model.py:47]: I don't get the difference between `if` and `else` here. The code seems to be DRYable.

File: openpype/hosts/blender/plugins/create/create_rig.py [hunk 5]
```diff
@@ -15,35 +15,58 @@ class CreateRig(plugin.Creator):
     family = "rig"
     icon = "wheelchair"
 
+    def get_selection_hierarchie(self):
```
Comment (Tilix4) [inline openpype/hosts/blender/plugins/create/create_rig.py:18]: `hierarchy`

File: openpype/hosts/blender/plugins/load/load_model.py [hunk 6]
```diff
@@ -41,68 +41,77 @@ def _remove(self, asset_group):
...
         ) as (data_from, data_to):
-            data_to.objects = data_from.objects
-
-        parent = bpy.context.scene.collection
+            data_to.collections = data_from.collections
 
-        empties = [obj for obj in data_to.objects if obj.type == 'EMPTY']
+        scene_collection = bpy.context.scene.collection
 
+        # Find the loaded collection and call it container
```
Comment (Tilix4) [inline openpype/hosts/blender/plugins/load/load_model.py:53]: I think it'd be better to rename `container` to `container_collection` to explicit the name

**Review by Tilix4 — COMMENTED** [review]
General comment: Your code is finely commented, that's nice! But pleas try to make sure comments start with a Cap `# My comment`, otherwise it looks like a commented code and not like an information.

File: openpype/hosts/blender/api/lib.py [hunk 7]
```diff
@@ -36,9 +33,10 @@ def register_module_call(mod):
             except:
                 traceback.print_exc()
         else:
-            print("\nWarning! '%s' has no register function, "
-                  "this is now a requirement for registerable scripts" %
-                  mod.__file__)
+            print(
+                "\nWarning! '%s' has no register function, "
```
Comment (Tilix4) [inline openpype/hosts/blender/api/lib.py:37]: Please try to use Python 3 fast strings all over your code: 
```suggestion
                f"\nWarning! '{mod.__file__}' has no register function, "
```

File: openpype/hosts/blender/api/lib.py [hunk 8]
```diff
@@ -130,7 +129,7 @@ def append_user_scripts():
     try:
         load_scripts(user_scripts.split(os.pathsep))
     except Exception:
-        print("Couldn't load user scripts \"{}\"".format(user_scripts))
+        print('Couldn\'t load user scripts "{}"'.format(user_scripts))
```
Comment (Tilix4) [inline openpype/hosts/blender/api/lib.py:132]: Same, use fstrings.

**Comment by Tilix4** [comment]
I'll keep updating encountered issues or requested changes here:


## Issues

- [ ] In Rigging step, when parenting objects to an armature, the `make local` is not done during this process. Therefore, if I publish my rig, the vertex groups and weight are lost. We must add a check in Pyblish blocking the process if it detects new vertex groups and the mesh data + object hasn't been made local.
  We could try to code that, but I'm not sure I support this solution, we must indicate what to do to the user, and in future, when the override system will be robust, we won't have to drop this piece of code.

## Cosmetics

- [ ] When creating an asset, if the original Collection becomes empty (because objects has been moved to the model's one), maybe we could delete this collection?

File: openpype/hosts/blender/api/plugin.py (hunk 1 above)
Comment (jojoba106) [reply openpype/hosts/blender/api/plugin.py:219]: It's not my function but they increment the name as long as it's in container_names. If it's not inside it's the right increment

**Review by Tilix4 — CHANGES_REQUESTED** [review]

**Comment by mkolar** [comment]
@jojoba106 @Tilix4 hey guys. Jus heads up that we've now managed to absorb all of the avalon-core codebase, so you'll need to merge develop into this and resolve any conflicts. Let us know if you need any help with that as it's quite the chunk of work you have here.

**Comment by antirotor** [comment]
Just pinging here about the state of things, is there any progress made or should we take over?

**Comment by BigRoy** [comment]
> It's been moved to https://github.com/ynput/OpenPype/pull/3463, we can close this one. And work is still being made for this very first big refactor, see https://github.com/ynput/OpenPype/discussions/3902
- [src](https://github.com/ynput/OpenPype/pull/2771#issuecomment-1341000786)

Closing this PR based on that comment.

**Pull request closed without merging**